### PR TITLE
SC-11420 Clipper for YFI-A

### PR DESCRIPTION
**Pull request by gbalabasquer**
# Description

Spell Description:

COB Team Name or Author(s):

# Contribution Checklist

- [ ] First commit title starts with COB Team Name and Collateral type (e.g. `___-TOKEN-X`)
- [ ] Code approved
- [ ] Tests approved
- [ ] CI Tests pass

# LIQ-2.0 Checklist

# `clipper` foreach `ilk`

- [ ] Validate constructor argument(s) are correct
- [ ] Ensure that the deployer was removed
- [ ] Ensure that the pause proxy was relied
- [ ] `dog.file(ilk, "clip", clipper)`
- [ ] `clipper.file("vow", vow)`
- [ ] `clipper.file("calc", calc)`
- [ ] `vat.rely(clipper)`
- [ ] `dog.rely(clipper)`
- [ ] `clipper.rely(dog)`
- [ ] `clipper.rely(end)`
- [ ] `clipper.rely(clipperMom)`
- [ ] `clipper.rely(esm)`
- [ ] `flipper.deny(cat)` (replaced by the clipper)
- [ ] `flipper.deny(flipperMom)` (to avoid reenabling the flipper without governance delay)
- [ ] `osm.kiss(clipper)`
- [ ] `osm.kiss(clipperMom)`
- [ ] `dog.file(ilk, "hole", "")`
- [ ] `dog.file(ilk, "chop", "")`
- [ ] `clipper.file("buf", "")`
- [ ] `clipper.file("tail", "")`
- [ ] `clipper.file("cusp", "")`
- [ ] `clipper.file("chip", "")`
- [ ] `clipper.file("tip", "")`
- [ ] Ensure that the deployer was removed from the `calc`
- [ ] Ensure that the pause proxy was relied on the `calc`
- [ ] `calc.file("step", "")`
- [ ] `calc.file("cut", "")`
- [ ] `clipperMom.setPriceTolerance(clipper, ???)`
- [ ] `clipper.upchost()`
- [ ] Replace flipper to clipper in the `ilk-registry`
- [ ] Add `clipper` and `calc` in the `chainlog`
- [ ] Remove `flipper` from the `chainlog`

# Checklist

- [ ] Every contract variable/method declared as public/external private/internal
- [ ] Verify expiration (`4 days + 2 hours` monthly and `30 days` for the rest)
- [ ] Verify hash in the description matches [here](https://emn178.github.io/online-tools/keccak_256.html)
- [ ] Validate all addresses used are in Kovan changelog
- [ ] Deploy spell to kovan `SOLC_FLAGS="--optimize --optimize-runs=1" dapp --use solc:0.6.11 build --extract && dapp create DssSpell --gas=5000000 --gas-price="$(seth --to-wei 3 "gwei")"`
- [ ] Ensure contract is verified on `kovan` etherscan
- [ ] Change test to use kovan spell address and deploy timestamp
- [ ] Keep `Kovan-DssSpell.sol` and `Kovan-DssSpell.t.sol` the same, but make a copy in `archive`
- [ ] `squash and merge` this PR

### CODE REVIEW

File: src/Kovan-DssSpell.sol [hunk 1]
```diff
@@ -33,110 +41,98 @@ contract DssSpellAction is DssAction {
...
+        // Whitelist CLIPPER_MOM in the YFI osm
+        DssExecLib.addReaderToOSMWhitelist(PIP_YFI, CLIPPER_MOM);
+
+        // No more auctions kicked via the CAT:
+        DssExecLib.deauthorize(MCD_FLIP_YFI_A, MCD_CAT);
+
+        // No more circuit breaker for the FLIP in YFI-A:
+        DssExecLib.deauthorize(MCD_FLIP_YFI_A, DssExecLib.flipperMom());
+
+        Fileable(MCD_DOG).file("YFI-A", "hole", 5_000 * RAD);
```
Comment (naszam) [inline src/Kovan-DssSpell.sol:103]: This should be 5M, maybe for kovan is fine 5k?

File: src/Kovan-DssSpell.t.sol [hunk 2]
```diff
@@ -1744,231 +1725,194 @@ contract DssSpellTest is DSTest, DSMath {
...
-
-        assertEq(end.wards(address(esm)), 1);
-        assertEq(vat.wards(address(esm)), 1);
-        assertEq(end_old.wards(address(esm_old)), 1);
-    }
+        assertEq(vat.wards(address(clipYFIA))    , 1);
+        assertEq(dog.wards(address(clipYFIA))    , 1);
+        assertEq(clipYFIA.wards(address(dog))    , 1);
+        assertEq(clipYFIA.wards(address(end))    , 1);
+        assertEq(clipYFIA.wards(address(clipMom)), 1);
```
Comment (naszam) [inline src/Kovan-DssSpell.t.sol:1759]:
```suggestion
        assertEq(clipYFIA.wards(address(clipMom)), 1);
        assertEq(clipYFIA.wards(address(esm)), 1);
```

File: src/Kovan-DssSpell.t.sol [hunk 3]
```diff
@@ -1744,231 +1725,194 @@ contract DssSpellTest is DSTest, DSMath {
...
 
-    function testSpellIsCast_new_End_functionality() public {
-        vote(address(spell));
-        scheduleWaitAndCast(address(spell));
-        assertTrue(spell.done());
+        assertEq(pipYFI.bud(address(clipYFIA)), 1);
```
Comment (naszam) [inline src/Kovan-DssSpell.t.sol:1761]:
```suggestion
        assertEq(pipYFI.bud(address(clipYFIA)), 1);
        assertEq(pipYFI.bud(address(clipMom)), 1);
```

**Review by naszam — COMMENTED** [review]
Checked:

- dss-Interfaces ok
- Clip YFI-A (verified, constructor args ok (vat, spotter, dog, ilk), wards ok, matches dss github)
- Calc YFI-A (verified, no constructor args, wards ok, matches dss github (abaci))
- LerpFab (verified, no constructor args, wards ok, flatten, matches github)
- YFI-A Liq 2.0 Params Ok (`hole` set to 5k Dai for kovan?)
- DssExecLib v0.0.6
- addresses_kovan.sol Matches
- Spell Actions Looks Good
- Tests Looks Good
- Test Pass:
```
dapp-test: rpc block: latest
Running 12 tests for src/Kovan-DssSpell.t.sol:DssSpellTest
[PASS] testOnTime() (gas: 1799008)
[PASS] testFailTooEarly() (gas: 10255)
[PASS] testLerp() (gas: 1938853)
[PASS] testClipperMomTripBreaker() (gas: 1887111)
[PASS] testSpellIsCast_GENERAL() (gas: 7230285)
[PASS] testSpellIsCast_YFI_A_CLIP() (gas: 2665729)
[PASS] testFailWrongDay() (gas: 10409)
[PASS] testCastCost() (gas: 1805473)
[PASS] testNewChainlogValues() (gas: 1829029)
[PASS] testClipperMomSetBreaker() (gas: 2014610)
[PASS] testFailTooLate() (gas: 10585)
[PASS] testFailClipperMomTripBreaker() (gas: 1828989)
```

Suggested (test PASS):

- Test to check clip wards for esm
- Test to check pip bud for clipMom

LGTM (to be deployed)

File: src/Kovan-DssSpell.sol (hunk 1 above)
Comment (gbalabasquer) [reply src/Kovan-DssSpell.sol:103]: Yes this is intentionally lower for kovan.

File: src/Kovan-DssSpell.t.sol [hunk 4]
```diff
@@ -1744,231 +1725,196 @@ contract DssSpellTest is DSTest, DSMath {
...
+            address(vat),
+            bytes32(uint256(keccak256(abi.encode(bytes32("YFI-A"), uint256(2)))) + 3),
+            bytes32(uint256(-1))
+        );
 
-        DSTokenAbstract BAT = DSTokenAbstract(addr.addr("BAT"));
-        GemJoinAbstract joinBATA = GemJoinAbstract(addr.addr("MCD_JOIN_BAT_A"));
-        FlipAbstract flipBATA = FlipAbstract(addr.addr("MCD_FLIP_BAT_A"));
+        // Add balance to the test address
+        uint256 ilkAmt = 0.05 * 10 ** 18;
```
Comment (hexonaut) [inline src/Kovan-DssSpell.t.sol:1773]: I don't normally see decimals. Could you assert that `ilkAmt > 0` just so I have confidence there isn't some rounding error which puts that to 0?

**Review by hexonaut — APPROVED** [review]
LGTM. One minor point in the tests. Looks good to deploy. Checked:

 * Contracts + wards good
 * Office hours off
 * Expiry is good
 * YFI convert to MIP45 good
 * Changelog updates good
 * Changelog version good
 * Tests look good
 * Tests passing

**Pull request merged**